pkg/cmd/root/help.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func rootHelpFunc(cmd *cobra.Command, _ []string) {
+	namePadding := 12
+	coreCmds := []string{}
+	configurationCmds := []string{}
+	libraryCmds := []string{}
+	infrastructureCmds := []string{}
+	additionalCmds := []string{}
+	for _, c := range cmd.Commands() {
+		if c.Short == "" {
+			continue
+		}
+		if c.Hidden {
+			continue
+		}
+		s := rpad(c.Name()+":", namePadding) + c.Short
+		if _, ok := c.Annotations[annotations.IsCore]; ok {
+			coreCmds = append(coreCmds, s)
+		} else if _, ok := c.Annotations[annotations.IsConfiguration]; ok {
+			configurationCmds = append(configurationCmds, s)
+		} else if _, ok := c.Annotations[annotations.IsLibrary]; ok {
+			libraryCmds = append(libraryCmds, s)
+		} else if _, ok := c.Annotations[annotations.IsInfrastructure]; ok {
+			infrastructureCmds = append(infrastructureCmds, s)
+		} else {
+			additionalCmds = append(additionalCmds, s)
+		}
+	}
+
+	// Assume all additional cmds are core if no core cmds are found
+	if len(coreCmds) == 0 {
+		coreCmds = additionalCmds
+		additionalCmds = []string{}
+	}
+
+	type helpEntry struct {
+		Title string
+		Body  string
+	}
+
+	longText := cmd.Long
+	if longText == "" {
+		longText = cmd.Short
+	}
+
+	helpEntries := []helpEntry{}
+	if longText != "" {
+		helpEntries = append(helpEntries, helpEntry{"", longText})
+	}
+	helpEntries = append(helpEntries, helpEntry{"USAGE", cmd.UseLine()})
+	if len(coreCmds) > 0 {
+		helpEntries = append(helpEntries, helpEntry{"CORE COMMANDS", strings.Join(coreCmds, "\n")})
+	}
+	if len(configurationCmds) > 0 {
+		helpEntries = append(helpEntries, helpEntry{"CONFIGURATION COMMANDS", strings.Join(configurationCmds, "\n")})
+	}
+	if len(libraryCmds) > 0 {
+		helpEntries = append(helpEntries, helpEntry{"LIBRARY COMMANDS", strings.Join(libraryCmds, "\n")})
+	}
+	if len(infrastructureCmds) > 0 {
+		helpEntries = append(helpEntries, helpEntry{"INFRASTRUCTURE COMMANDS", strings.Join(infrastructureCmds, "\n")})
+	}
+	if len(additionalCmds) > 0 {
+		helpEntries = append(helpEntries, helpEntry{"ADDITIONAL COMMANDS", strings.Join(additionalCmds, "\n")})
+	}
+
+	flagsUsage := cmd.LocalFlags().FlagUsages()
+	if flagsUsage != "" {
+		helpEntries = append(helpEntries, helpEntry{"FLAGS", dedent(flagsUsage)})
+	}
+	inheritedFlagUsage := cmd.InheritedFlags().FlagUsages()
+	if inheritedFlagUsage != "" {
+		helpEntries = append(helpEntries, helpEntry{"INHERITED FLAGS", dedent(inheritedFlagUsage)})
+	}
+
+	if !isRootCmd(cmd) {
+		if cmd.HasExample() {
+			helpEntries = append(helpEntries, helpEntry{"EXAMPLE", cmd.Example})
+		}
+	}
+
+	if isRootCmd(cmd) {
+		helpEntries = append(helpEntries, helpEntry{"LEARN MORE", `
+  Use "octopus [command] [subcommand] --help" for information about a command.`})
+		helpEntries = append(helpEntries, helpEntry{"FEEDBACK", "Open an issue: https://github.com/OctopusDeploy/cli/issues"})
+	}
+
+	out := cmd.OutOrStdout()
+	for _, e := range helpEntries {
+		if e.Title != "" {
+			fmt.Fprintln(out, output.Bold(e.Title))
```

Comment:
Super minor: rather than grabbing `cmd.OutOrStdout`, I've been calling `cmd.Printf` and friends (which are just wrappers that cobra has for you). Not sure if that's better or not 🤷